File: built-in plugins dsp/mb-brass-frenchhorn.cpp
/**
 * MB French Horn
 * Category : instrument
 * Type     : brass
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Majestic French horn with noble character
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_BRASS_FRENCHHORN_H
#define MB_BRASS_FRENCHHORN_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbBrassFrenchhorn : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-brass-frenchhorn";
    static constexpr const char* PLUGIN_NAME    = "MB French Horn";
    static constexpr const char* PLUGIN_TYPE    = "brass";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float warmth = 0.7f;  // range [0, 1]
    float vibrato = 0.25f;  // range [0, 1]
    float stopped = 0f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbBrassFrenchhorn() = default;
    ~MbBrassFrenchhorn() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.warmth = std::clamp(params.warmth, 0f, 1f);
        params.vibrato = std::clamp(params.vibrato, 0f, 1f);
        params.stopped = std::clamp(params.stopped, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB French Horn
        return input;
    }
};

#endif // MB_BRASS_FRENCHHORN_H

File: built-in plugins dsp/mb-brass-frenchhorn.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbBrassFrenchhornPlugin: PluginDefinition = {
  id: "mb-brass-frenchhorn",
  slug: "mb-brass-frenchhorn",
  name: "MB French Horn",
  category: "instrument",
  type: "brass" as any,
  version: "1.0.0",
  description: "Majestic French horn with noble character",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "sawtooth", detune: 0, gain: 0.4 },
    { type: "triangle", detune: 0, gain: 0.4 },
    { type: "sine", detune: -5, gain: 0.2 },
  ],
  envelope: { attack: 0.08, decay: 0.4, sustain: 0.75, release: 0.3 },
  parameters: [
    {
      id: "warmth",
      name: "Warmth",
      type: "float",
      defaultValue: 0.7,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "vibrato",
      name: "Vibrato",
      type: "float",
      defaultValue: 0.25,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "stopped",
      name: "Stopped",
      type: "float",
      defaultValue: 0,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { warmth: 0.7, vibrato: 0.25, stopped: 0, volume: 0.8 },
};

export default MbBrassFrenchhornPlugin;
